### PR TITLE
:rotating_light:(all): solve or ignore some pyrefly issues

### PR DESCRIPTION
**Pull request by o-laurent**
I've been mostly following copilot's suggestions. 

There may be better ways to handle the None model problem in the post_processing files. Do you have an opinion on this @alafage? I'm stopping there, still 382 errors, but the PR is going to be impossible to read otherwise.

- [x] Finish reducing the number of pyrefly errors
- [x] Fix coverage
- [ ] Squash merge

### CODE REVIEW

File: src/torch_uncertainty/datamodules/abstract.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def __init__(
         root: str | Path,
         batch_size: int,
         eval_batch_size: int | None,
-        val_split: float | None,
+        val_split: float | str | Path | None,
```
Comment (alafage) [inline src/torch_uncertainty/datamodules/abstract.py:39]: I think it's confusing to have str and Path types here. Why does pyrefly require them?

Comment (o-laurent) [reply src/torch_uncertainty/datamodules/abstract.py:39]: Because there's a possibility of providing a path to a file with lists of images to use a custom train/val/test split for some datasets. We should probably allow it for all of them though.

File: src/torch_uncertainty/datamodules/classification/imagenet.py
Comment (alafage) [inline src/torch_uncertainty/datamodules/classification/imagenet.py:56]: Shouldn't we have `val_split: float | str | Path | None = None` here?

Comment (o-laurent) [reply src/torch_uncertainty/datamodules/classification/imagenet.py:56]: We could add str, true!

File: src/torch_uncertainty/layers/functional/packed.py [hunk 2]
```diff
@@ -278,7 +278,10 @@ def packed_multi_head_attention_forward(  # noqa: D417
         Implementation of the packed multi-head attention is based on the PyTorch implementation of the
         `torch.nn.MultiheadAttention` module. The implementation is adapted to support packed inputs.
     """
-    is_batched = F._mha_shape_check(query, key, value, key_padding_mask, attn_mask, num_heads)
+    if hasattr(F, "_mha_shape_check"):
```
Comment (alafage) [inline src/torch_uncertainty/layers/functional/packed.py]: I'm not convinced here. F is something we import we shouldn't check whether the method exists

File: src/torch_uncertainty/layers/functional/packed.py [hunk 3]
```diff
@@ -336,7 +339,7 @@ def packed_multi_head_attention_forward(  # noqa: D417
     )
     if isinstance(embed_dim, Tensor):
         # embed_dim can be a tensor when JIT tracing
-        head_dim = embed_dim.div(num_heads, rounding_mode="trunc")
```
Comment (alafage) [inline src/torch_uncertainty/layers/functional/packed.py:339]: Wasn't the point to keep `head_dim` as a tensor for JIT Tracing? I'm not sure we can just fix that by converting it to an int

File: src/torch_uncertainty/layers/mc_batch_norm.py [hunk 4]
```diff
@@ -123,9 +123,9 @@ class MCBatchNorm1d(_MCBatchNorm):
         See `this issue/discussion <https://github.com/torch-uncertainty/torch-uncertainty/issues/218>`_.
     """
 
-    def _check_input_dim(self, inputs) -> None:
-        if inputs.dim() != 2 and inputs.dim() != 3:
-            raise ValueError(f"expected 2D or 3D input (got {inputs.dim()}D input)")
+    def _check_input_dim(self, input) -> None:  # pyrefly: ignore[bad-override] # noqa: A002
```
Comment (alafage) [inline src/torch_uncertainty/layers/mc_batch_norm.py:126]: Maybe we should stop using `input` and replace it with something like `x`

File: src/torch_uncertainty/losses/regression.py [hunk 5]
```diff
@@ -82,10 +82,15 @@ def _reg(self, dist: NormalInverseGamma | Independent, targets: Tensor) -> Tenso
 
     def forward(
         self,
-        dist: NormalInverseGamma,
+        dist: Distribution,
         targets: Tensor,
+        padding_mask: Tensor | None = None,
     ) -> Tensor:
-        loss_nll = super().forward(dist, targets)
+        if not isinstance(dist, NormalInverseGamma | Independent):  # coverage: ignore
```
Comment (alafage) [inline src/torch_uncertainty/losses/regression.py:89]: So Independent(Normal) would work? It feels like we don't cover all cases

File: src/torch_uncertainty/methods/deep_ensembles.py [hunk 6]
```diff
@@ -29,6 +32,7 @@ def forward(self, x: Tensor) -> Tensor:
                 where :math:`B` is the batch size, :math:`N` is the number of
                 estimators, and :math:`C` is the number of classes.
         """
+        # TODO: This probably doesn't work for regression
```
Comment (alafage) [inline src/torch_uncertainty/methods/deep_ensembles.py:35]: We should open an issue to keep track of that

File: src/torch_uncertainty/metrics/classification/brier_score.py [hunk 7]
```diff
@@ -110,7 +110,7 @@ def __init__(
             self.add_state("values", default=[], dist_reduce_fx="cat")
         self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
 
-    def update(self, probs: Tensor, target: Tensor) -> None:
+    def update(self, probs: Tensor, target: Tensor) -> None:  # pyrefly: ignore[bad-override]
```
Comment (alafage) [inline src/torch_uncertainty/metrics/classification/brier_score.py:113]: Is there a reason why we use `probs` instead of `preds` except for clarity?

**Review by alafage — COMMENTED** [review]
I think we need to rework on the parameter names in the metric's `update` functions to be sure the chosen naming is coherent and as close as possible as the ones in torchmetrics

File: src/torch_uncertainty/layers/functional/packed.py (hunk 2 above)
Comment (o-laurent) [reply src/torch_uncertainty/layers/functional/packed.py]: Unfortunately, it's not documented, so it will raise a pyrefly issue. Otherwise, we can just ignore it.

File: src/torch_uncertainty/layers/mc_batch_norm.py (hunk 4 above)
Comment (o-laurent) [reply src/torch_uncertainty/layers/mc_batch_norm.py:126]: Choosing input is needed for consistency due to the inheritance. Otherwise it will raise a pyrefly issue.

File: src/torch_uncertainty/methods/deep_ensembles.py (hunk 6 above)
Comment (o-laurent) [reply src/torch_uncertainty/methods/deep_ensembles.py:35]: There are a lot of TODOs and we probably should do this for all of them

File: src/torch_uncertainty/metrics/classification/brier_score.py (hunk 7 above)
Comment (o-laurent) [reply src/torch_uncertainty/metrics/classification/brier_score.py:113]: We can change for `preds`, right. In any case, there will be a pyrefly issue with the typings, I think.

File: src/torch_uncertainty/losses/regression.py (hunk 5 above)
Comment (o-laurent) [reply src/torch_uncertainty/losses/regression.py:89]: Keeping `Distribution` as a type is needed for consistency. In reality, DER expects only NormalInverseGamma distributions, and perhaps Independent[NormalInverseGamma]

File: src/torch_uncertainty/layers/functional/packed.py (hunk 3 above)
Comment (o-laurent) [reply src/torch_uncertainty/layers/functional/packed.py:339]: I thought that we could project it to an `int` there, but if you think that it might hurt performance, we can rollback.

Comment (alafage) [reply src/torch_uncertainty/layers/functional/packed.py:339]: I think JIT Tracing expects a tensor so it's better to roll back

Comment (o-laurent) [reply src/torch_uncertainty/layers/functional/packed.py:339]: Ok I'll do this nd convert the int to a tensor; with pyrefly, you cannot have same-named variables with different types if I understand correctly

File: src/torch_uncertainty/metrics/classification/brier_score.py (hunk 7 above)
Comment (o-laurent) [reply src/torch_uncertainty/metrics/classification/brier_score.py:113]: We will do this later, otherwise it will mess with the next PR.

File: src/torch_uncertainty/methods/deep_ensembles.py (hunk 6 above)
Comment (o-laurent) [reply src/torch_uncertainty/methods/deep_ensembles.py:35]: I created an issue for this one
